diagnostics: make paths to external items more visible

jseyfried opened the pull request:
This PR changes the reported path for an external item so that it is visible from at least one local module (i.e. it does not use any inaccessible external modules) if possible. If the external item's crate was declared with an `extern crate`, the path is guarenteed to use the `extern crate`.

Fixes #23224, fixes #23355, fixes #26635, fixes #27165.

r? @nrc

alexcrichton commented:
Holy cow I thought I'd never live to see the day this got implemented.

Thank you @jseyfried!

:cake:

mitaa commented:
Nice!

FYI, with #32293 there is also a refactoring going on here which lands soon. (https://github.com/rust-lang/rust/pull/32293/commits/684c734e603ce69480f1f56a48971df9f592bfbb, https://github.com/rust-lang/rust/pull/32293/commits/767c7897bd6a22994204619ccdba762550abd0ed, ...)

arielb1 commented on src/librustc_metadata/cstore.rs:
```diff
@@ -244,6 +247,52 @@ impl CStore {
         self.extern_mod_crate_map.borrow().get(&emod_id).cloned()
     }
+
+    pub fn visible_parent_map<'a>(&'a self) -> ::std::cell::RefMut<'a, DefIdMap<DefId>> {
```
Why the `RefMut`?

jseyfried replied:
The `RefCell` has to be borrowed mutably in this function and I didn't want to needlessly re-borrow it immutably before returning. This function is only used in `librustc_metadata` so I don't think it makes much of a difference.

nrc replied:
Could you document this function please?

jseyfried replied:
Will do

nrc commented on src/librustc_metadata/cstore.rs:
```diff
@@ -244,6 +247,52 @@ impl CStore {
+            };
+
+            let croot = DefId { krate: cnum, index: CRATE_DEF_INDEX };
+            for child in self.crate_top_level_items(cnum) { add_child(bfs_queue, child, croot); }
```
nit: spread over 3 lines

nrc commented on src/librustc_metadata/cstore.rs:
```diff
@@ -244,6 +247,52 @@ impl CStore {
+            let croot = DefId { krate: cnum, index: CRATE_DEF_INDEX };
+            for child in self.crate_top_level_items(cnum) { add_child(bfs_queue, child, croot); }
+            while let Some(def) = bfs_queue.pop_front() {
+                for child in self.item_children(def) { add_child(bfs_queue, child, def); }
```
and this one

nrc commented on src/librustc_metadata/csearch.rs:
```diff
@@ -140,6 +140,31 @@ impl<'tcx> CrateStore<'tcx> for cstore::CStore {
+    /// If possible, this returns a global path to the external item `def` that is visible from
+    /// at least one local module. If the external item's crate was declared with an
+    /// `extern crate`, the path is guarenteed to use the `extern crate`.
+    fn visible_item_path(&self, def: DefId) -> Option<Vec<hir_map::PathElem>> {
```
It might be simpler to return item_path() rather than None if there is no visible item path

jseyfried replied:
Good point

nrc commented:
r+ with the nits addressed

jseyfried commented:
@nrc thanks for the feedback! I addressed your comments in the most recent commit.

We should probably wait for #32293 to land before merging this since there will be non-trivial conflicts (cc @nikomatsakis).

bors commented:
:umbrella: The latest upstream changes (presumably #32293) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@jseyfried I think this can be r=nrc with `src/test/pretty/issue-4264.pp` updated to use the new paths (see travis failure).

jseyfried commented:
@bors r=nrc

bors commented:
:pushpin: Commit da41e58 has been approved by `nrc`

<!-- @bors r=nrc da41e583d614411c99370881c38f919c99dec448 -->

bors commented:
:hourglass: Testing commit da41e58 with merge 4583dc9...

bors commented:
:boom: Test timed out

eddyb commented:
@bors force retry

eddyb commented:
@bors p=1

alexcrichton commented:
This ended up passing all tests on bors, not sure what happened here so I'm just gonna merge manually
